pkg/kubelet/kuberuntime/kuberuntime_gc.go

```diff
@@ -0,0 +1,236 @@
+		}
+	}
+
+	if containerIndex != -1 {
+		if err := gc.client.RemoveContainer(containerID); err != nil {
+			glog.Warningf("Failed to remove dead container %q: %v", containerID, err)
+		}
+	}
+
+	return containerIndex
+}
+
+// removeOldestNContainers removes oldest n containers.
+func (gc *podGCInfo) removeOldestNContainers(n int) {
+	if n <= 0 {
+		return
+	}
+
+	toKeep := len(gc.containers) - n
+	for i := toKeep; i < len(gc.containers); i++ {
+		gc.removeContainer(gc.containers[i].containerID)
+	}
+
+	gc.containers = gc.containers[:toKeep]
+}
+
+// Newest first.
+type containerByCreated []*containerGCInfo
+
+func (a containerByCreated) Len() int           { return len(a) }
+func (a containerByCreated) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a containerByCreated) Less(i, j int) bool { return a[i].createdAt.After(a[j].createdAt) }
+
+// toContainerGCInfo converts kubecontainer.ContainerStatus to containerGCInfo.
+func toContainerGCInfo(podUID, sandboxID string, c *kubecontainer.ContainerStatus) *containerGCInfo {
+	return &containerGCInfo{
+		containerID: c.ID.ID,
+		createdAt:   c.CreatedAt,
+		sandboxID:   sandboxID,
+		podUID:      podUID,
+	}
+}
+
+// getEvictablePods gets evictable pods.
+// Evictable pods contain sandboxes and containers which are not active and created more than MinAge ago.
+func (m *kubeGenericRuntimeManager) getEvictablePods(minAge time.Duration) (map[string]*podGCInfo, error) {
+	sandboxes, err := m.getKubeletSandboxes(true)
+	if err != nil {
+		return nil, err
+	}
+
+	evictablePods := make(map[string]*podGCInfo)
+	newestGCTime := time.Now().Add(-minAge)
+	for _, s := range sandboxes {
+		if s.GetState() != runtimeApi.PodSandBoxState_NOTREADY {
+			continue
+		}
+
+		createdAt := time.Unix(*s.CreatedAt, 0)
+		if newestGCTime.Before(createdAt) {
+			continue
+		}
+
+		podUID := getStringValueFromLabel(s.Labels, types.KubernetesPodUIDLabel)
+		if len(podUID) == 0 {
+			podUID = nonKubernetesPodUID
+		}
+		if pod, ok := evictablePods[podUID]; ok {
+			pod.sandboxIDs = append(pod.sandboxIDs, s.GetId())
+		} else {
+			evictablePods[podUID] = &podGCInfo{
+				client:     m.runtimeService,
+				containers: make([]*containerGCInfo, 0),
+				sandboxIDs: []string{s.GetId()},
+			}
+		}
+
+		containerStatuses, err := m.getKubeletContainerStatuses(s.GetId())
+		if err != nil {
+			return nil, err
+		}
+		for _, c := range containerStatuses {
+			evictablePods[podUID].containers = append(evictablePods[podUID].containers,
+				toContainerGCInfo(podUID, s.GetId(), c))
+		}
+
+		sort.Sort(containerByCreated(evictablePods[podUID].containers))
+	}
+
+	return evictablePods, nil
+}
+
+// enforceMaxPerPodContainer enforces max number of dead containers of each pod and
+// returns the number of total evictable containers.
+func enforceMaxPerPodContainer(evictablePods map[string]*podGCInfo, maxPerPodContainer int) int {
+	totalEvictableContainers := 0
+	for _, pod := range evictablePods {
+		if maxPerPodContainer > 0 {
```

Comment:
The function is only used in this file, so either the caller should perform the check or the function should do it at the beginning. I think you are checking it in the loop because you want to calculate the `totalEvictableContainers` even if `maxPerPodContainer` is a negative value. I feel like this coupling may be too much. Worst case you'll calculate all the evictable containers in a separate step. That should not be too bad.